common/web/gesture-recognizer/src/engine/touchEventEngine.ts

```diff
@@ -115,49 +114,104 @@ export class TouchEventEngine<HoveredItemType, StateToken = any> extends InputEv
     // during a touchstart.)
     const allTouches = touchListToArray(event.touches);
     const newTouches = touchListToArray(event.changedTouches);
+    const oldTouches = allTouches.filter((touch1) => {
+      return newTouches.findIndex(touch2 => touch1.identifier == touch2.identifier) == -1;
+    });
 
-    this.eventDispatcher.runAsync(() => {
+    // Any 'old touches' should have pre-existing entries in our promise-map that are still current, as
+    // the promise-map is maintained 100% synchronously with incoming events.
+    const oldSourcePromises = oldTouches.map((touch) => this.pendingSourcePromises.get(touch.identifier));
+
+    this.eventDispatcher.runAsync(async () => {
+      const oldSources = await Promise.all(oldSourcePromises);
       // Maintain all touches in the `.touches` array that are NOT marked as `.changedTouches` (and therefore, new)
-      this.maintainTouchpointsWithIds(allTouches
-        .filter((touch1) => newTouches.findIndex(touch2 => touch1.identifier == touch2.identifier) == -1)
-        .map((touch) => touch.identifier)
-      );
+      this.maintainTouchpoints(oldSources);
+
+      return this.eventDispatcher.defaultWait;
     });
 
+    /*
+      We create Promises that can be set and retrieved synchronously with the actual event handlers
+      in order to prevent issues from tricky asynchronous identifier-to-source mapping attempts.
+
+      As these Promises are set (and thus, retrievable) synchronously with the actual event handlers,
+      we can closure-capture them for use in the internally-asynchronous processing closures.
+
+      `capturedSourcePromises` will be useful for closure-capture binding the new Promise(s) to
+      the closure to be queued.  `this.pendingSourcePromises` facilitates similar closure-capture
+      patterns within the touchMove and touchEnd handlers for their queued closures.
+    */
+    const capturedSourcePromises = new Map<number, ManagedPromise<GestureSource<ItemType, StateToken>>>();
+    for(let i=0; i < event.changedTouches.length; i++) {
+      const touch = event.changedTouches.item(i);
+      const promise = new ManagedPromise<GestureSource<ItemType, StateToken>>();
+      this.pendingSourcePromises.set(touch.identifier, promise);
+      capturedSourcePromises.set(touch.identifier, promise);
+    }
+
+    /*
+      When multiple touchpoints are active, we need to ensure a specific order of events.
+      The easiest way to ensure the exact order involves programmatic delay of their
+      processing, essentially "sequentializing" the events into a deterministic order.
```

Comment:
Is this delay visible to the end user?

Reply:
It should be imperceptible; it's a matter of queuing tasks against the microtask queue, the macrotask queue, and the custom queue from #10843.  Each of those should take negligible amounts of time.

The only delay added in this PR (at present) is due to the need to await a few pre-resolved Promises... so a couple of microtask delays per queued event should be the only extra wait here.  (They're not resolved when closure-captured, but they should already be resolved at every point they're being `await`-ed.)

Reply:
If I spent some time to dwell on things, perhaps I could 'rig' a way to avoid using Promises and the microtask queue for this, which would remove even the microtask queue waits... but that might actually make the code that little bit more difficult to parse for maintenance.  I feel that `Promise` semantics work in our favor here.

A simple object (providing a closure-capturable root reference) with a mutatable field would be enough.  Said field would start off `undefined` but be assigned the source's reference when it's available (during the final `onTouchStart` closure).  Of course, we'd have to explicitly document and maintain that connection, which is why I feel it a bit less "simple."

Reply:
My preference is almost always for synchronous code where possible -- re-entrant code makes code flow more complicated with potential for other interactions (from other promises and events) which all goes away without it. But I'd need to compare the two solutions here to be sure.